larreco/RecoAlg/ProjectionMatchingAlg.cxx

```diff
@@ -0,0 +1,1151 @@
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Class:       ProjectionMatchingAlg
+// Author:      D.Stefan (Dorota.Stefan@ncbj.gov.pl) and R.Sulej (Robert.Sulej@cern.ch), May 2015
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "larreco/RecoAlg/ProjectionMatchingAlg.h"
+
+#include "larcore/CoreUtils/ServiceUtil.h" // lar::providerFrom<>()
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "larreco/RecoAlg/PMAlg/Utilities.h"
+
+#include "larevt/CalibrationDBI/Interface/ChannelStatusService.h"
+#include "larevt/CalibrationDBI/Interface/ChannelStatusProvider.h"
+
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+pma::ProjectionMatchingAlg::ProjectionMatchingAlg(const fhicl::ParameterSet& pset)
+  : fDetProp(lar::providerFrom<detinfo::DetectorPropertiesService>())
+{
+	this->reconfigure(pset); 
+}
+// ------------------------------------------------------
+
+pma::ProjectionMatchingAlg::~ProjectionMatchingAlg(void)
+{
+}
+// ------------------------------------------------------
+
+void pma::ProjectionMatchingAlg::reconfigure(const fhicl::ParameterSet& p)
+{
+	fOptimizationEps = p.get< double >("OptimizationEps");
+	fFineTuningEps = p.get< double >("FineTuningEps");
+
+	fTrkValidationDist2D = p.get< double >("TrkValidationDist2D");
+	fHitTestingDist2D = p.get< double >("HitTestingDist2D");
+
+	fMinTwoViewFraction = p.get< double >("MinTwoViewFraction");
+	
+	fDetProp = lar::providerFrom<detinfo::DetectorPropertiesService>();
+	
+	pma::Node3D::SetMargin(p.get< double >("NodeMargin3D"));
+
+	pma::Element3D::SetOptFactor(geo::kZ, p.get< double >("HitWeightZ"));
+	pma::Element3D::SetOptFactor(geo::kV, p.get< double >("HitWeightV"));
+	pma::Element3D::SetOptFactor(geo::kU, p.get< double >("HitWeightU"));
+}
+// ------------------------------------------------------
+
+double pma::ProjectionMatchingAlg::validate(const pma::Track3D& trk,
+	const std::vector< art::Ptr<recob::Hit> >& hits,
+	unsigned int testView) const
+{
+	double max_d = fTrkValidationDist2D;
+	double d2, max_d2 = max_d * max_d;
+	unsigned int nAll = 0, nPassed = 0;
+
+	std::vector< unsigned int > trkTPCs = trk.TPCs();
+	std::vector< unsigned int > trkCryos = trk.Cryos();
+	std::map< std::pair< unsigned int, unsigned int >, std::pair< TVector2, TVector2 > > ranges;
```

Comment:
This is my first attempt at commenting - please let me know if you see it.

1)
LArSoft should make a global decision about standard 2-vector, 3-vector, 4-vector classes (and the classes that perform transformations on them).  When making this decision, be aware that a an object of type TVector2 is also a TObject and has a size of 32 bytes - 16 bytes of (x,y) and 16 bytes of TObject-ness.

2)
In this line there are two uses of std::pair. I think that this code would be easier to understand if these are replaced by two well named structs each with two well named elements.  

3)
It would be a good idea if the geometry subsystem provided a 0-based dense integer identifier for all of the tpc+cryo in the system.  Then the std::map could be replaced by a std::vector.